### PR TITLE
Support while / do-while loops; desugar general C-style for (#26)

### LINKED ISSUE #26: Support while / do-while loops (incompatible with @total)
Third feature of the subset-widening roadmap.

**Fails today:**

```ts
function leftPad(str: string, len: number, ch: string): string {
  let pad = str;
  while (pad.length < len) {   // TH0010: Loop not supported; use recursion or array methods
    pad = ch + pad;
  }
  return pad;
}
```

**After:** accepted; emits Lean do-notation `while` (do-while via `repeat ... until`). JS loops can diverge, so partial-backed emission is honest semantics.

**Policy:** Lean backs do-notation `while` with a `partial` combinator — fine for evaluation (byte-match contract unaffected), opaque to proofs. `@total` + `while` gets a clean TH diagnostic ("use for-of or recursion, or drop @total"), mirroring the @throws/@total exclusivity design.

Also closes the general C-style `for` lowering left open in #25. Depends on #24.

Metric: test262 `language/statements/{while,do-while}/` slices (needs #23).

### PR DESCRIPTION
**Pull request by jessealama**
Closes #26. Fourth feature of the subset-widening roadmap; depends on #24/#25 (merged).

## What's admitted now

- **`while (test) body`** — any test expression; lowers to Lean do-notation `while`. Lean's `while` re-checks the condition after `continue`, matching TS, so unlabeled `break`/`continue` map 1:1.
- **`do body while (test)`** — lowers to `repeat ... until !(test)`; the body runs at least once, as in TS. A **loop-level `continue` stays rejected** (TH0010): TS `continue` jumps to the test, but `repeat ... until` re-enters the body without checking it — the lowering would diverge where TS exits.
- **General C-style `for`** (closes the lowering left open in #25) — desugared at the AST level to `init; while (test) { body; update }`, reusing the while path. Loop-level `continue` with an update clause stays rejected (the desugared body would skip the update). The canonical `for (let i = 0; i < B; i++)` shape keeps its structural `[0:B]` range lowering and never falls back to the desugar, so the #25 operand boundaries (string-length bounds, bound mutation, var reassignment) are unchanged.

Loops remain function-scoped (module-level loops are #49) and excluded from `@throws`/try contexts, as in #25.

## `@total` × loops: new TH0068

Both lowerings are backed by a partial combinator, which the lake-side termination verification (TH0070) cannot see through — `@total` + `while` would verify vacuously. Per the issue's policy, the combination draws a clean diagnostic instead (TH0068, `totalHasUnverifiableLoop`), mirroring the `@total`/`@throws` exclusivity. `for-of`/canonical-for stay `@total`-friendly (pinned by `loop-total-forof.ts` and a new unit test).

## Runtime

The left-pad milestone fixture needs string concatenation, which had never worked (`a + b` on strings emitted `HAdd String String` with no instance — the pure path included). One runtime instance lands with the example, per convention: `HAdd String String String := ⟨String.append⟩`. Mixed string/number `+` deliberately keeps no instance — loud Lean-stage failure rather than a miscompile.

## Harness: bucket↔outcome enforcement (TDD red)

`run-examples.js` never checked a pass label against its bucket, so an accept/ fixture that thales subset-rejects passed as `ok (subset-rejected)` — the new fixtures couldn't fail first, and a regression demoting an accepted example would have gone unnoticed. Each bucket now requires its one pass label (accept→accepted, mirror→tsc-rejected, reject→subset-rejected, throws→both-throw). The four new accept fixtures were verified RED under this check (bucket-mismatch) before the feature landed, plus the TH0068 fixture (directive-check).

## Fixtures

- `accept/loop-while-left-pad.ts` — the issue's left-pad example, byte-matched through Lean
- `accept/loop-do-while-runs-once.ts` — body-runs-once semantics on a false-on-entry test
- `accept/loop-while-break-continue.ts` — `while (true)` with break/continue
- `accept/loop-for-general-step.ts` — non-canonical for (non-zero start, `>`, `i -= 2`)
- `reject/0068-total-while-loop.ts`, `reject/0010-do-while-continue.ts`
- `reject/0010-for-noncanonical.ts` repurposed to the still-rejected continue-with-update shape; `0010-for-loop.ts` recommented as the module-level guard

## Verification

`lake build` + `lake build ThalesTest` clean (LoopEscape/LoopSubsetCheck/LoopShape/LeanSyntaxDo tests extended to pin the new behavior); `format:check` clean; harness self-test 18/0; full conformance **120/0**. test262 re-measure reproduces the #25 table byte-identically — expected, not an artifact: the slice tests are top-level statement scripts, so TH0010 movement there is gated entirely on #49 (documented in `docs/test262.md`).

### CODE REVIEW

**Review by jessealama — COMMENTED** [review]
Review focus: comment hygiene, duplication, branching, and structural drift. No correctness bugs found — the lowering semantics, escape analysis, and fixture coverage all verified clean (including the repurposed reject fixtures: the now-accepted shapes are covered by the four new accept fixtures and the extended Lean unit tests).

**Cross-cutting: issue-number references in comments and docs.** This PR adds ~45 new `#26` / `#25/#26` references in comments across source (`EscapeAnalysis.lean`, `Lean.lean`, `LeanSyntax.lean`, `LoopShape.lean`, `SubsetCheck.lean`), all four `Test/Emit` test files, and docs (`errors.md`, `subset.md`, `subset` tables). Issue numbers shouldn't be visible to programmers — comments should speak for themselves. The good news is the comments already contain the real explanation after the number, so the fix is mostly deleting the `#26:` prefixes. This continues a pre-existing `#24`/`#25` pattern, so a full cleanup is a repo-wide sweep; at minimum this PR shouldn't add new ones. (`docs/test262.md` is arguably exempt — it's a roadmap-metrics doc structurally keyed to issue numbers.)

Inline comments below cover the rest: a silent-skip hazard in the new harness enforcement, the twice-written for-desugar decomposition, two checker/emitter drift hazards, a thrice-repeated guard cascade, and two over-narrating comment blocks.

File: scripts/run-examples.js [hunk 1]
```diff
@@ -406,6 +406,18 @@ function collectCases(rootDir, mode) {
   return cases;
 }
 
+/** Directory membership IS the test specification: each bucket admits
+ *  exactly one pass label. A pass outcome with the wrong label (e.g. an
+ *  accept/ file that is merely subset-rejected) is a corpus violation, not
+ *  a pass — without this, a regression that demotes an accepted example to
+ *  subset-rejected would go unnoticed. */
+const expectedLabelByBucket = {
```
Comment (jessealama) [inline scripts/run-examples.js]: **Two sources of truth, and a silent-skip hazard.** `expectedLabelByBucket` duplicates the bucket list already defined at line 379 (`buckets = ['accept', 'mirror', 'reject', 'throws']`). If a future bucket is added to one but not the other, `expected` comes back `undefined` at line 435, and `if (expected && ...)` silently disables the bucket↔outcome enforcement for that bucket — the exact failure mode (pass-with-wrong-label) this check was built to catch.

Suggest a single bucket→label table that drives both the directory walk and the enforcement, so an unmapped bucket is impossible by construction (or at least throw on a missing mapping instead of skipping).

File: Thales/Emit/Lean.lean [hunk 2]
```diff
@@ -1206,7 +1206,52 @@ partial def emitBodyDo (env : EmitEnv) (info : EscapeAnalysis.MutationInfo)
...
+      -- routing for the init declarator). A missing test is `while true`.
+      -- The init binding outliving the loop in the Lean block is safe:
+      -- shadowing rejection (TH0032) keeps any same-named outer binding
+      -- out, so no later read can resolve to the loop variable.
+      | .notLowerable =>
+          match s with
+          | .forStmt fb init test update body =>
+              if LoopShape.generalForDesugarable s
+                  && !(LoopShape.hasLabeledBreakOrContinue body) then
+                let initStmts : List Statement := match init with
```
Comment (jessealama) [inline Thales/Emit/Lean.lean]: **The for→while desugar decomposition is written out twice.** This `initStmts`/`testE`/`bodyStmts ++ update` extraction is mirrored by parallel match arms in `SubsetCheck.lean` (general-for arm, ~346–356), which destructures the same `init`/`test`/`update` to check the would-be-desugared parts. The checker and emitter can drift into checking and emitting *different* desugarings.

A shared `LoopShape.desugarGeneralFor : Statement → Option (initStmts × test × bodyStmts)` would let both call sites consume one decomposition, making disagreement impossible by construction. Canonical-for and for-of avoid this hazard by never desugaring — the duplication is new with this PR.

File: Thales/Emit/SubsetCheck.lean [hunk 3]
```diff
@@ -291,25 +296,64 @@ partial def checkStmt (ctx : MutCtx) (stmt : Statement) : Array Diagnostic :=
...
+      match LoopShape.classifyLoop s with
+      | .canonicalFor _ (.inl _) _ => true
+      | .canonicalFor _ (.inr arrName) _ => identIsArray ctx arrName
+      | _ => false
+    if loopContextAdmitted ctx && canonicalAdmitted then
       -- Only the body: classifyLoop already constrained init/test/update to
       -- bare shapes, and routing `i++` through checkExpr would draw
       -- .assignmentInExpressionPosition.
       checkStmt ctx body
+    else if loopContextAdmitted ctx && LoopShape.generalForDesugarable s then
```
Comment (jessealama) [inline Thales/Emit/SubsetCheck.lean]: **Admission predicate is weaker than the emitter's.** This arm checks `generalForDesugarable s` but not `hasLabeledBreakOrContinue body`; the emitter (`Lean.lean:1219–1220`) requires both. For e.g. `for (let i = 5; i > 0; i--) { outer: { break outer; } }`, this guard alone would admit what the emitter refuses — today only EscapeAnalysis's poisoning (`hasUnloweredLoopShape` → `loopContextAdmitted = false`) keeps TH0010 firing. That makes the admission condition split across two files that must stay in sync.

Mirroring the emitter's exact condition here (or sharing one predicate between the two) closes the gap.

File: Thales/Emit/SubsetCheck.lean [hunk 4]
```diff
@@ -291,25 +296,64 @@ partial def checkStmt (ctx : MutCtx) (stmt : Statement) : Array Diagnostic :=
...
+  -- lake-side termination verification pass vacuously.
+  | .whileStmt b test body =>
+    if loopContextAdmitted ctx then
+      if ctx.inTotalFn then
+        #[mkThalesDiag .totalHasUnverifiableLoop b.loc]
+      else
+        checkExpr ctx test ++ checkStmt ctx body
+    else
+      #[mkThalesDiag .loopNotSupported b.loc]
+  | .doWhileStmt b body test =>
```
Comment (jessealama) [inline Thales/Emit/SubsetCheck.lean:324]: **No SubsetCheck-level shape guard for do-while + loop-level `continue`.** Unlike the `forStmt`/`forOfStmt` arms, which call `LoopShape` predicates directly, this arm gates only on `loopContextAdmitted` — the TH0010 rejection for do-while-with-continue lives entirely in EscapeAnalysis (`hasOwnUnlabeledContinue` poisoning, `EscapeAnalysis.lean:282–288`). If that detection is ever refactored away, SubsetCheck accepts silently and only the emitter's defence-in-depth re-check stands between that and a divergent lowering, with no diagnostic-level backstop.

A direct `LoopShape.hasOwnUnlabeledContinue body` check here would make the rejection self-contained, consistent with how the for shapes anchor theirs.

File: Thales/Emit/SubsetCheck.lean [hunk 5]
```diff
@@ -291,25 +296,64 @@ partial def checkStmt (ctx : MutCtx) (stmt : Statement) : Array Diagnostic :=
...
+  -- admitted-shape loop and an unlowerable one (labeled break/continue,
+  -- do-while with loop-level `continue`) rejects BOTH loops.
+  --
+  -- #26: while/do-while have no per-loop shape conditions beyond what
+  -- EscapeAnalysis already folded into `doModeLowerable` (labels,
+  -- do-while `continue`). In a `@total` function the admitted shapes draw
+  -- TH0068 instead: their lowering is partial-backed and would make the
+  -- lake-side termination verification pass vacuously.
+  | .whileStmt b test body =>
+    if loopContextAdmitted ctx then
```
Comment (jessealama) [inline Thales/Emit/SubsetCheck.lean]: **Same guard cascade three times.** The `if loopContextAdmitted ctx → if ctx.inTotalFn → TH0068 → else recurse → else TH0010` shape repeats verbatim here, in the `doWhileStmt` arm just below, and (modulo the outer guard) in the general-for arm at ~343. The next loop form adds a fourth copy, and any refinement of the `@total`×loop policy has to be edited in three places.

A small helper taking `(ctx, loc, bodyChecks)` would collapse these, in line with the file's existing helper style (`loopContextAdmitted`, `routeIdentMutation`).

File: Thales/Emit/Lean.lean [hunk 6]
```diff
@@ -1206,7 +1206,52 @@ partial def emitBodyDo (env : EmitEnv) (info : EscapeAnalysis.MutationInfo)
...
+                let bodyStmts := blockStmts body
+                  ++ (match update with
+                      | some u => [.exprStmt fb u]
+                      | none => [])
+                emitBodyDo env info
+                  (initStmts
+                    ++ .whileStmt fb testE (.blockStmt fb bodyStmts) :: rest)
+              else unloweredDoStmt
+          | _ => unloweredDoStmt
+  -- #26 loop lowering: `while (c) body` → `while c do …`; `do body while (c)`
```
Comment (jessealama) [inline Thales/Emit/Lean.lean]: **Comment over-narrates.** Six lines, most of it reviewer-directed architecture rationale ("defence-in-depth against phase drift", "like the for cases above") restating what EscapeAnalysis already did upstream — and the phase-drift rationale is already stated once at the #25 comment block this points to. The load-bearing content is one line: TS loops *while* the test holds, Lean's `repeat ... until` loops *until* it does. Suggest trimming to that constraint.

File: Thales/Emit/Lean.lean [hunk 7]
```diff
@@ -1206,7 +1206,52 @@ partial def emitBodyDo (env : EmitEnv) (info : EscapeAnalysis.MutationInfo)
                 { env with bindingEnv := env.bindingEnv.insert i .number }
               .forDo i iterExpr (shim :: emitBodyDo env' info (blockStmts body))
                 :: emitBodyDo env info rest
-      | .notLowerable => unloweredDoStmt
+      -- #26: a non-canonical `for` desugars at the AST level to
```
Comment (jessealama) [inline Thales/Emit/Lean.lean]: **Comment conflates three concerns.** This block mixes (a) flow narration — "re-enters this function, reusing the while lowering" — which the code below already shows, (b) the genuine constraint that the init binding outliving the loop is safe *only because* TH0032 shadowing rejection exists, and (c) the missing-test-defaults-to-`true` rule, which belongs at the match arm that implements it. Suggest keeping only the shadowing constraint here and moving/dropping the rest.

File: scripts/run-examples.js (hunk 1 above)
Comment (jessealama) [reply scripts/run-examples.js]: Addressed in 85b90f8e: a single `corpusBuckets` table now drives both the corpus walk (`Object.keys(corpusBuckets)`) and the outcome enforcement, and the `expected &&` guard is gone — an unmapped bucket now fails loudly as `bucket-mismatch` instead of silently skipping the check.

File: Thales/Emit/Lean.lean (hunk 2 above)
Comment (jessealama) [reply Thales/Emit/Lean.lean]: Addressed in 85b90f8e: the decomposition now lives once in `LoopShape.desugarGeneralFor`, and EscapeAnalysis, SubsetCheck, and the emitter all consume it — the phases can no longer disagree about whether a `for` desugars or what it desugars to. New `LoopShapeTest` cases pin the shape, the missing-test → `while (true)` default, and the labeled-break → `none` condition.

File: Thales/Emit/SubsetCheck.lean (hunk 3 above)
Comment (jessealama) [reply Thales/Emit/SubsetCheck.lean]: Addressed in 85b90f8e: the `hasLabeledBreakOrContinue` condition is folded into the shared `LoopShape.desugarGeneralFor`, which this arm now consumes — checker and emitter apply the same admission predicate by construction, and SubsetCheck checks the exact statement list the emitter will lower.

File: Thales/Emit/SubsetCheck.lean (hunk 4 above)
Comment (jessealama) [reply Thales/Emit/SubsetCheck.lean:324]: Addressed in 85b90f8e: the `doWhileStmt` arm now checks `LoopShape.hasOwnUnlabeledContinue body` directly, so the TH0010 rejection is anchored in SubsetCheck itself rather than relying solely on EscapeAnalysis poisoning.

File: Thales/Emit/SubsetCheck.lean (hunk 5 above)
Comment (jessealama) [reply Thales/Emit/SubsetCheck.lean]: Addressed in 85b90f8e: the cascade is now one `checkAdmittedLoop` helper used by the while and do-while arms; the general-for copy disappeared entirely, since that arm now checks the desugared statements and the while arm supplies the `@total` gate (same diagnostic, same source location).

File: Thales/Emit/Lean.lean (hunk 6 above)
Comment (jessealama) [reply Thales/Emit/Lean.lean]: Addressed in 85b90f8e: trimmed to three lines centered on the load-bearing constraint — TS loops WHILE the test holds, Lean's `repeat` runs UNTIL it does — with a one-clause pointer to the defence-in-depth convention established above.

File: Thales/Emit/Lean.lean (hunk 7 above)
Comment (jessealama) [reply Thales/Emit/Lean.lean]: Addressed in 85b90f8e: the flow narration is reduced to two lines at the call site, and the TH0032 shadowing-safety constraint and missing-test default moved into `LoopShape.desugarGeneralFor`'s docstring, where the desugaring now lives.

**Comment by jessealama** [comment]
Review findings addressed in 85b90f8e, including the cross-cutting one from the review body: all issue-number references this branch introduced are removed from source comments, test banners/error strings, and `docs/errors.md` / `docs/subset.md` (the comments now lead with the semantic reason that previously followed the number). Pre-existing references from earlier work are left for a repo-wide sweep, and `docs/test262.md` keeps its references as the roadmap-metrics doc.

Verification after the refactor: `lake build` + `lake build ThalesTest` clean, harness self-test 18/0, conformance 120/0, `format:check` clean.
